File: 0001-0500/0159_lengthOfLongestSubstringTwoDistinct.py
class Solution:
    def lengthOfLongestSubstringTwoDistinct(self, s):
        window = {}
        left = 0
        longest = 0
        for right in range(len(s)):
            if len(window) < 2:
                window[s[right]] = window.get(s[right], 0) + 1
                longest = max(longest, right - left + 1)
            elif s[right] in window:
                window[s[right]] += 1
                longest = max(longest, right - left + 1)
            else:
                while len(window) == 2:
                    window[s[left]] -= 1
                    if window[s[left]] == 0:
                        del window[s[left]]
                    left += 1
                window[s[right]] = 1
        return longest

# Alternative solution

class Solution:
    def lengthOfLongestSubstringTwoDistinct(self, s):
        window = {}
        longest = 0
        i = j = 0
        
        while i < len(s):
            while len(window) == 2 and s[i] not in window:
                window[s[j]] -= 1
                
                if not window[s[j]]:
                    del window[s[j]]
                
                j += 1
                    
            window[s[i]] = window.get(s[i], 0) + 1
            longest = max(longest, sum(window.values()))
            i += 1
        
        return longest

# Alternative solution

class Solution:
    def lengthOfLongestSubstringTwoDistinct(self, s):
        window = {}
        start = 0
        answer = 0
        
        for end in range(len(s)):
            window[s[end]] = window.get(s[end], 0) + 1
            
            while len(window) > 2:
                window[s[start]] -= 1
                if window[s[start]] == 0:
                    del window[s[start]]
                start += 1
                
            answer = max(answer, end - start + 1)
        
        return answer

File: 0001-0500/0159_lengthOfLongestSubstringTwoDistinct.java
class Solution {
    public int lengthOfLongestSubstringTwoDistinct(String s) {
        HashMap<Character, Integer> hashmap = new HashMap<>();
        int start = 0;
        int longest = 0;
        
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            
            while (hashmap.size() == 2 && !hashmap.containsKey(c)) {
                hashmap.put(s.charAt(start), hashmap.get(s.charAt(start)) - 1);
                if (hashmap.get(s.charAt(start)) == 0) {
                    hashmap.remove(s.charAt(start));
                }
                start++;
            }
            
            hashmap.put(c, hashmap.getOrDefault(c, 0) + 1);
            longest = Math.max(longest, i - start + 1);
        }
        
        return longest;
    }
}
